packages/langium/src/workspace/document-builder.ts

```diff
@@ -352,14 +352,23 @@ export class DefaultDocumentBuilder implements DocumentBuilder {
 
     protected async runCancelable(documents: LangiumDocument[], targetState: DocumentState, cancelToken: CancellationToken,
         callback: (document: LangiumDocument) => MaybePromise<unknown>): Promise<void> {
-        const filtered = documents.filter(e => e.state < targetState);
+        const filtered = documents.filter(doc => doc.state < targetState);
         for (const document of filtered) {
             await interruptAndCheck(cancelToken);
             await callback(document);
             document.state = targetState;
-            await this.notifyDocumentPhase(document, targetState, cancelToken);
         }
-        await this.notifyBuildPhase(filtered, targetState, cancelToken);
+
+        // notify when document state equals target state
+        // because initial state for parsed documents is Parsed.
+        for (const doc of documents) {
+            const docs = [];
+            if (doc.state === targetState) {
+                await this.notifyDocumentPhase(doc, targetState, cancelToken);
+                docs.push(doc);
+            }
+            await this.notifyBuildPhase(docs, targetState, cancelToken);
+        }
```

Comment:
This is not the intended behavior:
1. `notifyDocumentPhase` is supposed to run immediately after the document has been processed - that's the whole Raison d'être (see also the documentation on `DocumentBuilder#onDocumentPhase`). Otherwise we will run into https://github.com/eclipse-langium/langium/discussions/1562 again.
2. I'm a bit confused by putting `notifyBuildPhase` and the `docs` list inside the loop. Shouldn't they be outside of it?

Reply:
Regarding 1. Didn't know that. I will move it back
Regarding 2. Sorry, that should be outside the loop. 
First update ⬇️ More tomorrow